### PR TITLE
Implemented `repeat` option

### PR DESCRIPTION
**Pull request by stephandesouza**
Allow Paratest to run multiple times, as `--repeat` present on PHPUnit

image

I could not find a better way to:

- [ ] avoid incremental test count
- [ ] avoid the `Single-run is self-completed` part

### CODE REVIEW

**Comment by Slamdunk** [comment]
Instead of running the suite 3 times, what about duplicating the suites/files *n* times, here https://github.com/paratestphp/paratest/blob/2fd895d4e255a612a2d323185e93f46d6cef6fb5/src/Runners/PHPUnit/SuiteLoader.php#L168-L172 ?

**Comment by stephandesouza** [comment]
> Instead of running the suite 3 times, what about duplicating the suites/files _n_ times, here
> 
> https://github.com/paratestphp/paratest/blob/2fd895d4e255a612a2d323185e93f46d6cef6fb5/src/Runners/PHPUnit/SuiteLoader.php#L168-L172
> 
> ?

I need to apply some tests, but I guess this will not work when filtering the tests.

**Comment by Slamdunk** [comment]
>  I guess this will not work when filtering the tests.

Why not?

**Comment by stephandesouza** [comment]
> > I guess this will not work when filtering the tests.
> 
> Why not?

Suites are indexed by its filename/suitename on `initSuites()`, I guess, to prevent miss duplications.

Running with 1 (default) repeat, focus on config.
image

Running part with 3 repeat, focus on proposed change
image

**Comment by Slamdunk** [comment]
And what about simply `--passthru` it like other options like `--group`:

https://github.com/paratestphp/paratest/blob/2fd895d4e255a612a2d323185e93f46d6cef6fb5/src/Runners/PHPUnit/Options.php#L417

```
$ vendor/bin/paratest
Running phpunit in 4 processes with /var/www/html/vendor/phpunit/phpunit/phpunit                                                                                                                                                                
                                                                                                                                                                                                                                                
Configuration read from /var/www/html/phpunit.xml                                                                                                                                                                                               
                                                                                                                                                                                                                                                
....................                                              20 / 20 (100%)

Time: 00:09.778, Memory: 16.00 MB

OK (20 tests, 39 assertions)


$ vendor/bin/paratest --passthru='--repeat 3'
Running phpunit in 4 processes with /var/www/html/vendor/phpunit/phpunit/phpunit

Configuration read from /var/www/html/phpunit.xml

............................................................      60 / 60 (100%)

Time: 00:00.560, Memory: 16.00 MB

OK (60 tests, 117 assertions)
```

**Comment by stephandesouza** [comment]
> And what about simply `--passthru` it like other options like `--group`:
This simplifies a lot! I will make the rollback and adapt the `Option` class

File: phpstan-baseline.neon [hunk 1]
```diff
@@ -94,6 +94,11 @@ parameters:
 			count: 1
 			path: src/Runners/PHPUnit/Options.php
 
+		-
+			message: "#^Parameter \\#37 \\$repeat of class ParaTest\\\\Runners\\\\PHPUnit\\\\Options constructor expects string\\|null, array\\<string\\>\\|bool\\|int\\|string\\|null given\\.$#"
```
Comment (Slamdunk) [inline phpstan-baseline.neon:98]: This shouldn't be needed

File: src/Runners/PHPUnit/Options.php [hunk 2]
```diff
@@ -219,6 +219,8 @@ final class Options
     private $orderBy;
     /** @var int */
     private $randomOrderSeed;
+    /** @var string|null */
```
Comment (Slamdunk) [inline src/Runners/PHPUnit/Options.php:222]:
```suggestion
    /** @var int */
```

File: src/Runners/PHPUnit/Options.php [hunk 3]
```diff
@@ -264,7 +266,8 @@ private function __construct(
         int $verbosity,
         ?string $whitelist,
         string $orderBy,
-        int $randomOrderSeed
+        int $randomOrderSeed,
+        ?string $repeat
```
Comment (Slamdunk) [inline src/Runners/PHPUnit/Options.php:270]:
```suggestion
        int $repeat
```

File: src/Runners/PHPUnit/Options.php [hunk 4]
```diff
@@ -425,6 +430,17 @@ public static function fromConsoleInput(InputInterface $input, string $cwd): sel
...
 
+        if (is_string($options['repeat'])) {
+            if (! is_numeric($options['repeat'])) {
+                throw new InvalidArgumentException(sprintf(
+                    'Option --repeat should have a number value, "%s" given',
+                    $options['repeat']
+                ));
+            }
+
+            $filtered['repeat'] = $options['repeat'];
```
Comment (Slamdunk) [inline src/Runners/PHPUnit/Options.php:441]:
```suggestion
            if ($options['repeat'] !== (string) (int) $options['repeat']) {
                throw new InvalidArgumentException(sprintf(
                    'Option --repeat should have an integer number value, "%s" given',
                    $options['repeat']
                ));
            }

            $filtered['repeat'] = (int) $options['repeat'];
```

File: src/Runners/PHPUnit/Options.php [hunk 5]
```diff
@@ -536,7 +552,8 @@ public static function fromConsoleInput(InputInterface $input, string $cwd): sel
             $verbosity,
             $options['whitelist'],
             $options['order-by'] ?? self::ORDER_DEFAULT,
-            (int) $options['random-order-seed']
+            (int) $options['random-order-seed'],
+            $options['repeat']
```
Comment (Slamdunk) [inline src/Runners/PHPUnit/Options.php:556]:
```suggestion
            (int) $options['repeat']
```

File: src/Runners/PHPUnit/Options.php [hunk 6]
```diff
@@ -739,6 +756,12 @@ public static function setInputDefinition(InputDefinition $inputDefinition): voi
                 InputOption::VALUE_OPTIONAL,
                 'Use a specific random seed <N> for random order'
             ),
+            new InputOption(
+                'repeat',
+                'r',
```
Comment (Slamdunk) [inline src/Runners/PHPUnit/Options.php:761]: I prefer no shortcuts, like PHPUnit:
```suggestion
                null,
```

File: src/Runners/PHPUnit/Options.php [hunk 7]
```diff
@@ -1085,6 +1108,11 @@ public function randomOrderSeed(): int
         return $this->randomOrderSeed;
     }
 
+    public function repeat(): ?string
```
Comment (Slamdunk) [inline src/Runners/PHPUnit/Options.php:1111]:
```suggestion
    public function repeat(): int
```

File: test/Unit/Runners/PHPUnit/OptionsTest.php [hunk 8]
```diff
@@ -296,6 +304,7 @@ public function testProvidedOptions(): void
             '--whitelist' => 'WHITELIST',
             '--order-by' => Options::ORDER_RANDOM,
             '--random-order-seed' => (string) $expected_random_seed,
+            '--repeat' => '2',
```
Comment (Slamdunk) [inline test/Unit/Runners/PHPUnit/OptionsTest.php:307]:
```suggestion
            '--repeat' => 2,
```

File: test/Unit/Runners/PHPUnit/OptionsTest.php [hunk 9]
```diff
@@ -333,6 +342,7 @@ public function testProvidedOptions(): void
         static::assertSame('WHITELIST', $options->whitelist());
         static::assertSame(Options::ORDER_RANDOM, $options->orderBy());
         static::assertSame($expected_random_seed, $options->randomOrderSeed());
+        static::assertSame('2', $options->repeat());
```
Comment (Slamdunk) [inline test/Unit/Runners/PHPUnit/OptionsTest.php:345]:
```suggestion
        static::assertSame(2, $options->repeat());
```

**Review by Slamdunk — CHANGES_REQUESTED** [review]

File: src/Runners/PHPUnit/Options.php (hunk 2 above)
Comment (stephandesouza) [reply src/Runners/PHPUnit/Options.php:222]: Will we enforce always sending the repeat option to PHPUnit? If so, we need to set default to '1', no?

File: src/Runners/PHPUnit/Options.php (hunk 6 above)
Comment (stephandesouza) [reply src/Runners/PHPUnit/Options.php:761]: Ooops, okay.

File: src/Runners/PHPUnit/Options.php (hunk 2 above)
Comment (Slamdunk) [reply src/Runners/PHPUnit/Options.php:222]: No: the option in passed only within the `if {}` with the  `$filters` variable.
The `Options::$repeat` attribute is here only as a reference of the CLI arguments

File: src/Runners/PHPUnit/Options.php (hunk 4 above)
Comment (stephandesouza) [reply src/Runners/PHPUnit/Options.php:441]: `$filtered['repeat'] = (int) $options['repeat'];`

will request a chage on $filtered signature:

`* @param array<string, string|int|null> $filtered`

Okay on this?

File: src/Runners/PHPUnit/Options.php (hunk 2 above)
Comment (stephandesouza) [reply src/Runners/PHPUnit/Options.php:222]: Oh, okay.

File: src/Runners/PHPUnit/Options.php (hunk 4 above)
Comment (Slamdunk) [reply src/Runners/PHPUnit/Options.php:441]: Sorry, didn't check that: ok for `$filtered['repeat'] = $options['repeat'];` then.

**Review by Slamdunk — APPROVED** [review]

**Pull request merged**